Graph Equivalence

emmanvg opened the pull request:
This PR provides the ability for a user to perform equivalence checks for two graphs.

- Moves semantic equivalence related functions to their own module. Added extra imports in `environment.py` to prevent breaking change
- Available via `stix2.environment.Environment.graphically_equivalent()` but also via `stix2.equivalence.graph_equivalence.graphically_equivalent()`
- Optional ability to perform versioned check when the are multiple versions of the same object present on the provided graph
- Configurable depth to prevent excessive recursion
- Ideally the complete graphs are provided - failure to de-reference object results in zero for the property contributing score
- More objects and property checks are available in Graph Based Semantic Equivalence

emmanvg committed:
new packages for graph and object-based semantic equivalence

emmanvg committed:
new method graphically_equivalent for Environment, move equivalence methods out

emmanvg committed:
object equivalence function, methods used for object-based moved here.

emmanvg committed:
new graph_equivalence methods

emmanvg committed:
add notes

emmanvg committed:
add support for versioning checks (default disabled)

emmanvg committed:
new tests to cover graph equivalence and new methods

emmanvg committed:
added more imports to environment.py to prevent breaking changes

emmanvg committed:
variable changes, new fields for checks, reset depth check per call

emmanvg committed:
flexibility when object is not available on graph.

emmanvg committed:
refactor debug logging message

emmanvg committed:
new file stix2.equivalence.graph_equivalence.rst and stix2.equivalence.object_equivalence.rst for docs

emmanvg committed:
API documentation for new modules

emmanvg committed:
additional text required to build docs

emmanvg committed:
add more test methods for list_semantic_check an graphically_equivalent/versioning

emmanvg committed:
add logging debug messages, code clean-up

emmanvg committed:
include individual scoring on results dict, fix issue on list_semantic_check not keeping highest score

emmanvg committed:
include results as summary in prop_scores, minor tweaks

emmanvg committed:
Update __init__.py

doctrings update

clenk commented on stix2/equivalence/graph_equivalence/__init__.py:
```diff
@@ -0,0 +1,209 @@
+        "name": (30, partial_string_based),
+        "external_references": (60, partial_external_reference_based),
+        "created_by_ref": (5, semantic_check),
+        "object_marking_refs": (5, list_semantic_check),
```
As discussed offline, let's remove `created_by_ref` and `object_marking_refs` from the calculation since two organizations might create what is semantically the same content, or that same content might get marked differently by different organizations.

clenk commented on stix2/equivalence/object_equivalence/__init__.py:
```diff
@@ -0,0 +1,434 @@
+    return result
+
+
+def semantic_check(ref1, ref2, ds1, ds2, **weights):
```
It seems like this function is checking de-referenced objects rather than semantics in general. Maybe rename to `reference_check` or `dereferenced_check`? Same with `list_semantic_check`.

clenk commented on stix2/equivalence/object_equivalence/__init__.py:
```diff
@@ -0,0 +1,434 @@
+def list_semantic_check(refs1, refs2, ds1, ds2, **weights):
+    """For objects that contain multiple references (i.e., object_refs) perform
+    the same de-reference for object-based semantic equivalence. The score influences
+    the objects containing all references. The result is weighted on the longest ref list"""
```
Missing a period.

clenk reviewed (commented):
Looking good!

Can we make the names of the semantic equivalence submodules consistent? Right now we have `graph_equivalence`, `object_equivalence`, and `patterns`. Since they're already under `equivalence` it's redundant to include it at this level too I think.

emmanvg committed:
apply feedback from pull request

- rename semantic_check to reference_check
- rename modules to graph and object respectively to eliminate redundancy
- remove created_by_ref and object_marking_refs from graph WEIGHTS and rebalance

emmanvg committed:
update docs/ entries

emmanvg committed:
add more checks, make max score based on actual objects checked instead of the full list, only create entry when type is present in WEIGHTS dictionary

update tests to reflect changes

emmanvg committed:
rename package patterns -> pattern

emmanvg committed:
documentation, moving weights around

emmanvg committed:
more documentation moving

clenk commented on stix2/equivalence/graph/__init__.py:
```diff
@@ -95,88 +100,36 @@ def graphically_equivalent(ds1, ds2, prop_scores={}, **weight_dict):
-        "tdelta": 1,  # One day interval
-    },
-    "intrusion-set": {
+WEIGHTS.update({
```
Should we copy `WEIGHTS` and update/use the copy instead? That way if someone performs graph equivalence and then later performs object equivalence it won't do the additional stuff that graph equivalence does.

emmanvg replied:
I think the fact that they also shared the same name caused a problem in scope resolution. I renamed the graph one...

emmanvg committed:
rename WEIGHTS variable for graph_equivalence

clenk reviewed (approved):
LGTM, thanks @emmanvg!
